lib/Dialect/LLHD/IR/LLHDOps.cpp

```diff
@@ -388,6 +388,66 @@ OpFoldResult llhd::ShrOp::fold(ArrayRef<Attribute> operands) {
       });
 }
 
+//===----------------------------------------------------------------------===//
+// ExtractElementOp
+//===----------------------------------------------------------------------===//
+
+OpFoldResult llhd::ExtractElementOp::fold(ArrayRef<Attribute> operands) {
+  uint64_t index = indexAttr().getInt();
+
+  // llhd.extract_element(llhd.shr(hidden, base, constant amt), index)
+  if (auto shrOp = target().getDefiningOp<llhd::ShrOp>()) {
+    IntegerAttr intAttr;
+    if (matchPattern(shrOp.amount(), m_Constant<IntegerAttr>(&intAttr))) {
+      uint64_t amt = intAttr.getValue().getZExtValue();
+
+      // with amt + index < baseWidth
+      //   => llhd.extract_element(base, amt + index)
+      if (amt + index < shrOp.getBaseWidth()) {
+        targetMutable().assign(shrOp.base());
+        setAttr("index", IntegerAttr::get(indexAttr().getType(), amt + index));
+        return result();
+      }
+
+      // with amt + index >= baseWidth && amt + index < baseWidth + hiddenWidth
+      //   => llhd.extract_element(hidden, amt + index - baseWidth)
+      if (amt + index < shrOp.getBaseWidth() + shrOp.getHiddenWidth()) {
+        targetMutable().assign(shrOp.hidden());
+        setAttr("index", IntegerAttr::get(indexAttr().getType(),
+                                          amt + index - shrOp.getBaseWidth()));
+        return result();
+      }
+    }
+  }
+
+  // llhd.extract_element(llhd.array(a_0, ..., a_n), i) => a_i
+  if (auto arrayOp = target().getDefiningOp<llhd::ArrayOp>()) {
+    uint64_t index = indexAttr().getValue().getZExtValue();
+    if (index < arrayOp.values().size()) {
```

Comment:
I'm not sure if this is possible, but it looks like this is something that the verifier for the operation could catch.  If that is the case, it would be better to check it there than to have defensive code throughout the codebase -- one omission would be a bug.

Reply:
Currently, this is indeed checked by the verifier, but there was some discussion recently about whether the static extract operations should do bound checks or allow out-of-bound extracts by returning a default value to make them more similar to the dynamic extract operations. If the latter is the case then this check here would be necessary. So it's defensive in regard to what will happen with the extract operation itself. Maybe @fabianschuiki wants to comment on this?

Reply:
I think we want out-of-bounds accesses to be well-defined also for the static `extract_element` operation. The rationale is that you frequently have something like the following in SystemVerilog:
```systemverilog
module bootrom (input logic [2:0] addr, output logic [31:0] data);
    localparam logic [31:0] rom [5] = '{ ... }; /* 5 entries */
    assign data <= rom[addr]; /* can address 8 entries */
endmodule
```
In SystemVerilog, out-of-bounds accesses are well-defined: addresses 0 to 4 would produce values from the `rom`, addresses 5 to 7 would produce `'x` (or `'0` if `data` were of type `bit [31:0]`). Synthesizers frequently use `x` values to better optimize boolean logic, like the `rom` lookup above. The hardware generated for `rom[addr]` above would produce a value for all of the 8 possible addresses, just with the upper 3 being garbage. I think it's quite crucial that we reflect this behaviour for the `extract_element` and `dynamic_extract_element` corresponding to the `rom[addr]`, since LLHD lives at the very low end of the abstraction stack.

What do you guys think?

By the way, it gets even crazier for dynamic part-selects, which essentially are shifts without sign/zero-extension:
```systemverilog
module dummyrom (input logic [2:0] addr, output logic [3:0] data);
    localparam logic [4:0] rom = 5'b10101;
    assign data = rom[addr +: 4];
endmodule
// Produces the following output:
// addr  output
// 0     0101
// 1     1010
// 2     X101
// 3     XX10
// 4     XXX1
// 5     XXXX
// 7     XXXX
// 8     XXXX
```

Reply:
I'm not sure what the right semantics are here for LLHD - I'm just suggesting that you make a call and then enforce it consistently.  Unneeded defensive code is just dead code that leads to bugs and confusion.

It would be totally reasonable to allow out of bound dynamic array indices, but I don't see why you'd ever allow that for tuples.

Reply:
> I'm not sure what the right semantics are here for LLHD - I'm just suggesting that you make a call and then enforce it consistently. Unneeded defensive code is just dead code that leads to bugs and confusion.

Yeah good point.

> It would be totally reasonable to allow out of bound dynamic array indices, but I don't see why you'd ever allow that for tuples.

Totally agree; array indices and slices are what I had in mind. Not sure how you would even determine the result type of a dynamic field extract on a struct. I also don't know of any HDLs that allow for something like that.

Reply:
Sorry for the long pause on my side.
That's a good point indeed. I removed the bounds checks in the folder as they are currently enforced in the verifier.

Thanks @fabianschuiki for this detailed explanation. Maybe we should add this as an issue to keep track of it? I agree that these semantics would make more sense for LLHD.